src/ott/neural/solvers/base_solver.py

```diff
@@ -0,0 +1,480 @@
+# Copyright OTT-JAX
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+from pathlib import Path
+from types import MappingProxyType
+from typing import Any, Callable, Dict, Literal, Mapping, Optional, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+import optax
+from flax.training import train_state
+from jax import random
+
+from ott.geometry import costs, pointcloud
+from ott.geometry.pointcloud import PointCloud
+from ott.neural.models import models
+from ott.problems.linear import linear_problem
+from ott.problems.quadratic import quadratic_problem
+from ott.solvers.linear import sinkhorn
+
+
+class BaseNeuralSolver(ABC):
+  """Base class for neural solvers.
+
+  Args:
+    iterations: Number of iterations to train for.
+    valid_freq: Frequency at which to run validation.
+  """
+
+  def __init__(self, iterations: int, valid_freq: int, **_: Any) -> None:
+    self.iterations = iterations
+    self.valid_freq = valid_freq
+
+  @abstractmethod
+  def setup(self, *args: Any, **kwargs: Any) -> None:
+    """Setup the model."""
+    pass
+
+  @abstractmethod
+  def __call__(self, *args: Any, **kwargs: Any) -> None:
+    """Train the model."""
+    pass
+
+  @abstractmethod
+  def transport(self, *args: Any, forward: bool, **kwargs: Any) -> Any:
```

Comment:
isn't it more coherent if the solver's state or output can transport?

Reply:
Yes, we were thinking about this, i.e. whether we should have an output class or not. For now, we thought it might be better to just have the `transport` method as a method of the solver class, as it allows to interrupt and continue training straightforwardly. Moreover, we thought it might be more coherent with loading and saving the model. To be precise, if we had an output class, we would load a model into a state, which then is passed to the model, similarly when saving a model, we would load the model to an output class, which we would then save. It seemed a little bit like a detour. 

But happy to change, if you prefer.